### PR TITLE
[Profiler] Improve reference chain 

### PR DESCRIPTION
**Pull request by chrisnas**
## Summary of changes
- Add namespace to type name in reference chain
- Avoid stopping the generation of class histogram on error

## Reason for change
Help comparing reference chain with class histogram

## Implementation details
- change the serialization to integrate namespaces
- improve error management to avoid cutting the chain and allow next computation

## Test coverage
- update tests
- add new tests

### CODE REVIEW

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    auto entry = _resolvedTypes.find(typeName);
+    if (entry != _resolvedTypes.end())
+    {
+        return entry->second;
+    }
+
+    ClassID classId = ResolveTypeNoLock(moduleId, fullTypeName);
+
+    // Failures are cached too: a type missing from the core library will never show up later.
+    _resolvedTypes[typeName] = classId;
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp]: you can use std::move(typeName) here

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    Log::Debug("Core library module found: ", assemblyName, " (module ", moduleId, ")");
+    return true;
+}
+
+ModuleID CoreLibModuleProvider::GetModuleId() const
+{
+    return _moduleId.load(std::memory_order_acquire);
+}
+
+bool CoreLibModuleProvider::TryGetMetadata(IMetaDataImport2** ppMetadataImport)
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp]: Based on the usage, this could 
 ComPtr<IMetaDataImport2> CoreLibModuleProvider::GetMetadata()

And then check for nullity.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp:121]: you may want to add a mechanism to check that the runner actually started 
something like https://github.com/DataDog/dd-trace-dotnet/pull/9037/changes#diff-612ba090bb73a099eca085ef02218dd0f8bd10555b77655dd7b01be635731340R127

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp [hunk 3]
```diff
@@ -30,55 +39,213 @@ const InlineVTCache::InlineVTInfo* InlineVTCache::GetInlineVTInfo(ClassID classI
...
-    return insertedIt->second.has_value() ? &insertedIt->second.value() : nullptr;
+    // Inspecting the type here would mean resolving its fields types while the runtime is
+    // suspended for the dump: this is done later, once the dump is over.
+    _pendingClassIDs.insert(classID);
+
+    // Record a "nothing to attribute" placeholder so the remaining objects of this type
+    // stop at the cache lookup above instead of probing the queue again, once per object,
+    // while the runtime is suspended. ResolvePendingTypes drops it before inspecting the
+    // type for real.
+    _cache.emplace(classID, std::nullopt);
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp:71]: This can lead to a crash:
this function returns a pointer to a cell of the unordered_map. If the `cache` is rehash (caused by emplace), this pointer will point to some places that might crash at some point.

do not fill the cache here, it will be done later in `ResolvePendingTypes` when walking the `_pendingClassIDs` 

The first dump, we will have misses but this should be ok, right ?
A test can be added (if possible):
Cache type A with a field of type B. Hold A’s InlineVTInfo* (or use A.fields), then call GetInlineVTInfo(B) many times / with enough distinct misses to force a rehash. With the old code that can crash or corrupt; with the fix, A’s data is still intact.

**Review by gleocadie — APPROVED** [review]
# medium findings

## 1. `GetClassLayout` on every GC ref of every inline VT

**Today:** On later dumps, for every object whose type is already in `InlineVTCache`, every GC pointer goes through `GetInlineValueTypeOwner`. That function calls `ICorProfilerInfo::GetClassLayout` **once per candidate inline-VT field** just to get `vtSize` for a range check. Nested VTs do it again. That is a profiler API call on the dump-GC thread, under the fault guard.

Size is a property of the `ClassID`. `BuildInlineVTInfo` already calls `GetClassLayout` after the dump, but only for the **containing** type, and it throws `classSize` away. Nested VT size is never stored.

The struct only keeps offset + ClassID:

```50:53:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.h
    struct InlineVTInfo
    {
        std::vector<std::pair<ULONG, ClassID>> fields;
    };
```

Build stores the pair, no size:

```257:293:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
    ULONG fieldCount = 0;
    ULONG classSize = 0;
    HRESULT hr = _pCorProfilerInfo->GetClassLayout(
        classID, nullptr, 0, &fieldCount, &classSize);
    // ...
    if (fieldInfo.isValueType && fieldInfo.valueTypeClassID != 0)
    {
        inlineVtFields.emplace_back(fieldInfo.offset, fieldInfo.valueTypeClassID);
    }
```

Hot loop during dump:

```515:550:profiler/src/ProfilerEngine/Datadog.Profiler.Native/ReferenceChainTraverser.cpp
    for (const auto& [vtOffset, vtClassID] : vtInfo.fields)
    {
        ULONG vtStart = baseOffset + vtOffset;
        ULONG fieldCount = 0;
        ULONG vtSize = 0;
        HRESULT hr = _pCorProfilerInfo->GetClassLayout(vtClassID, nullptr, 0, &fieldCount, &vtSize);
        if (FAILED(hr) || vtSize == 0)
        {
            continue;
        }
        if (refOffset < vtStart || refOffset >= vtStart + vtSize)
        {
            continue;
        }
```

How often: **per object** with cached inline VTs → **per GC ref** → **per inline-VT field** (plus nesting). First dump for a type does not hit this (`GetInlineVTInfo` returns `nullptr`).

**What to do:** Store size at inspection time and do arithmetic during the dump.

```cpp
// InlineVTCache.h
struct InlineVTField
{
    ULONG offset = 0;
    ClassID classID = 0;
    ULONG size = 0;
};
struct InlineVTInfo
{
    std::vector<InlineVTField> fields;
};
```

In `BuildInlineVTInfo`, when recording a VT field, `GetClassLayout` on `fieldInfo.valueTypeClassID` and store `vtSize` (skip if 0). Then in `GetInlineValueTypeOwner`:

```cpp
for (const auto& field : vtInfo.fields)
{
    ULONG vtStart = baseOffset + field.offset;
    if (refOffset < vtStart || refOffset >= vtStart + field.size) continue;
    // ...
}
```

Also switch `AddInlineValueTypeInstances` (`field.second` → `field.classID`), `MergeParentInlineVTs`, and `GetMemorySize`.

---

## 2. Failed CoreLib resolutions cached forever

**Today:** Once CoreLib’s module ID is known, `ResolveTypeInCoreLib` writes **every** result into `_resolvedTypes`, including `ClassID == 0`. `GetModuleMetaData` failure, a missing type, and `GetClassFromTokenAndTypeArgs` failure all look the same (`0`). A one-time metadata miss then poisons that type name for the process.

If CoreLib is **not loaded yet** (`_moduleId == 0`), it returns 0 **without** caching. That part is fine.

```71:98:profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp
    ClassID classId = ResolveTypeNoLock(moduleId, fullTypeName);
    // Failures are cached too: a type missing from the core library will never show up later.
    _resolvedTypes[typeName] = classId;
    return classId;
```

`0` from metadata not ready vs type missing:

```101:147:profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp
    IMetaDataImport2* pMetadataImport = GetMetadataNoLock();
    if (pMetadataImport == nullptr) return 0; // GetModuleMetaData failed
    if (FAILED(FindTypeDefByName) || not TypeDef) return 0;
    return SUCCEEDED(GetClassFromTokenAndTypeArgs) ? classId : 0;
```

Only production caller is `InlineVTCache::ResolvePrimitiveClassID` (generic-arg primitives like `KeyValuePair<int, object>`). `InlineVTCache` itself does **not** cache `0` in `_primitiveClassIDs`, but the next lookup still goes to CoreLib and hits the cached 0.

The unit test currently **requires** failures to be cached (`CoreLibModuleProviderTest.cpp` around the second `System.Int32` lookup expecting no extra `GetModuleMetaData`).

**What to do:** Cache successes only:

```cpp
ClassID classId = ResolveTypeNoLock(moduleId, fullTypeName);
if (classId != 0)
{
    _resolvedTypes[typeName] = classId;
}
return classId;
```

CoreLib names are a tiny set; retries are cheap. Flip the existing test so a failed lookup may call `GetModuleMetaData` again. Add a test: first resolve fails (no metadata), then metadata becomes available, second resolve returns non-zero.

Stricter option: still cache a genuine `FindTypeDefByName` miss, but never cache when `GetMetadataNoLock() == nullptr`. The simple “cache only non-zero” is enough.

---

## 3. Heap-range filter removed

**Today:** `GCHeapRangeSet` was deleted in `6325d8e0ee`. The only remaining check is null / low / alignment. Any other aligned garbage pointer is treated as an object and read under `MemoryFaultGuard`.

```613:626:profiler/src/ProfilerEngine/Datadog.Profiler.Native/ReferenceChainTraverser.cpp
bool ReferenceChainTraverser::IsValidObjectAddress(uintptr_t address)
{
    if (address == 0 || address < 0x10000)
    {
        return false;
    }
    if ((address % sizeof(void*)) != 0)
    {
        return false;
    }
    return true;
}
```

Used on every discovered ref in `DrainTraversalStack` and VT-array refs.

On Linux, `EnsureInstalled` registers SIGSEGV/SIGBUS and **ignores** whether `RegisterHandler` succeeded:

```119:134:profiler/src/ProfilerEngine/Datadog.Profiler.Native/MemoryFaultGuard.h
    auto* segv = ProfilerSignalManager::Get(SIGSEGV);
    if (segv != nullptr)
    {
        segv->RegisterHandler(&details::FaultHandler);
    }
```

If that fails, `Run()` still `sigsetjmp`s; the next fault is a real crash. Contrast: `TimerCreateCpuProfiler` does check `RegisterHandler`.

The old filter was accept-biased: seed from `GetGenerationBounds` **before** the dump GC, grow from validated roots, empty set = accept everything. Header said a false accept costs one recovered fault; a false reject silently drops a real edge.

**Crashtracker nuance:** in this repo, Linux crash tracking is createdump-based; it does not `sigaction` SIGSEGV. CPU sampling uses SIGUSR1/SIGPROF. `FaultHandler` returns `false` when not guarded so the previous (CLR) handler still runs. The “someone else steals SIGSEGV later” risk is real in general, but there is no crashtracker SIGSEGV installer on Linux here. The bigger issue is: no range filter, and a failed handler install is silent.

**What to do:**

1. Restore `GCHeapRangeSet` from `6325d8e0ee^`: seed in `StartGCDump` via `GetGenerationBounds` (illegal during dump GC, so it has to be before the EventPipe session), grow from validated roots in `OnBulkRoot*`, check it at the end of `IsValidObjectAddress`. Keep fail-open (empty set accepts everything). Keep `MemoryFaultGuard` for the false accepts.

2. Log if registration fails:

```cpp
if (segv == nullptr || !segv->RegisterHandler(&details::FaultHandler))
{
    Log::Warn("MemoryFaultGuard: failed to register SIGSEGV handler; "
              "heap-snapshot traversal cannot recover from access faults.");
}
// same for SIGBUS
```

The guard stays; the filter is so typical junk never takes a signal.

#  low / nit findings

## 1. `WasAbortedByFaults()` is also set for C++ exceptions (Low)

**Today:** A C++ exception that escapes `TraverseFromSingleRootCore` is caught and routed to `OnTraversalAborted()`, which sets `_faultBudgetExhausted = true` without incrementing `_faultCount`. `WasAbortedByFaults()` is just that flag. `HeapSnapshotManager::OnEndGCDump` then treats it as a fault-budget abort: it logs “cut short by 0 memory access fault(s)” and increments `_consecutiveFaultyDumps`, which can permanently disable traversal after 3 dumps.

Real SIGSEGV/SIGBUS recovery goes through `OnTraversalFault()`, which does increment `_faultCount` and only then may set the same flag. Memory faults never hit `OnTraversalAborted`. C++ exceptions (`std::bad_alloc`, CLR exceptions from profiling APIs) propagate out of the guard by design and hit this catch. A unit test already asserts `WasAbortedByFaults() == true` with `GetFaultCount() == 0`.

```42:55:profiler/src/ProfilerEngine/Datadog.Profiler.Native/ReferenceChainTraverser.cpp
void ReferenceChainTraverser::TraverseFromSingleRoot(const RootInfo& root)
{
    try
    {
        TraverseFromSingleRootCore(root);
    }
    catch (...)
    {
        OnTraversalAborted();
    }
```

```161:176:profiler/src/ProfilerEngine/Datadog.Profiler.Native/ReferenceChainTraverser.cpp
void ReferenceChainTraverser::OnTraversalAborted()
{
    _faultBudgetExhausted = true;
    LogOnce(Warn,
        "Reference-chain traversal was aborted by an unexpected exception. "
        "Traversal is skipped for the rest of this heap snapshot. "
        "The class histogram is unaffected.");
}
```

```70:73:profiler/src/ProfilerEngine/Datadog.Profiler.Native/ReferenceChainTraverser.h
    bool WasAbortedByFaults() const { return _faultBudgetExhausted; }
```

```138:148:profiler/src/ProfilerEngine/Datadog.Profiler.Native/ReferenceChainTraverser.cpp
void ReferenceChainTraverser::OnTraversalFault()
{
    _faultCount++;
    if (_faultCount >= MaxFaultsPerDump)
    {
        _faultBudgetExhausted = true;
    }
```

```756:786:profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp
    else if (_pReferenceChainTraverser->WasAbortedByFaults())
    {
        _consecutiveFaultyDumps++;
        Log::Warn("Reference-chain traversal was cut short by ", _lastTraversalFaultCount,
            " memory access fault(s) (", _consecutiveFaultyDumps,
            " consecutive dump(s)). The reference tree for this dump is partial.");

        if (_consecutiveFaultyDumps >= MaxConsecutiveFaultyDumps)
        {
            _gcDescDisabled = true;
            Log::Warn("Reference-chain traversal has been disabled for the remainder of the process "
                "after ", MaxConsecutiveFaultyDumps, " consecutive dumps hit the memory access "
                "fault budget. Heap class histograms are unaffected.");
        }
    }
```

**Why it matters:** One throw (e.g. `bad_alloc`) is logged as zero memory-access faults and counts toward the 3-dump permanent disable. The traverser’s own log is accurate (“aborted by an unexpected exception”); the manager’s policy and log are not.

**What to do:** Add a separate abort flag (e.g. `_abortedByException`) set only in `OnTraversalAborted()`. Keep `WasAbortedByFaults()` as `_faultBudgetExhausted && _faultCount >= MaxFaultsPerDump` (or `_faultCount > 0`). In `OnEndGCDump`, branch: exception abort → log the exception and do **not** increment `_consecutiveFaultyDumps`; fault-budget abort → keep the current consecutive-dump policy. Update `ExceptionEscapingTheGuardAbortsTheDumpWithoutCountingAFault` to assert the new exception-abort API, not `WasAbortedByFaults()`.

---

## 2. `EnsureInstalled` ignores `RegisterHandler` failure (Low)

**Today:** On Linux, `EnsureInstalled` registers `FaultHandler` for SIGSEGV and SIGBUS via `ProfilerSignalManager`, but discards `RegisterHandler`’s `bool`. `call_once` never retries. `Run()` still `sigsetjmp`s and arms `t_inGuardedRegion`. If registration failed, the handler is not installed, so a fault in a guarded region is not recovered — the previous/default handler runs and the process dies. Windows is a no-op (SEH needs no registration). Contrast: `TimerCreateCpuProfiler` checks the return and refuses to start if registration failed.

```116:136:profiler/src/ProfilerEngine/Datadog.Profiler.Native/MemoryFaultGuard.h
inline void EnsureInstalled()
{
#ifndef _WINDOWS
    static std::once_flag installed;
    std::call_once(installed, [] {
        auto* segv = ProfilerSignalManager::Get(SIGSEGV);
        if (segv != nullptr)
        {
            segv->RegisterHandler(&details::FaultHandler);
        }
        auto* bus = ProfilerSignalManager::Get(SIGBUS);
        if (bus != nullptr)
        {
            bus->RegisterHandler(&details::FaultHandler);
        }
    });
#endif
}
```

```50:75:profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/ProfilerSignalManager.cpp
bool ProfilerSignalManager::RegisterHandler(HandlerFn_t handler)
{
    _isHandlerInPlace = SetupSignalHandler();
    if (_isHandlerInPlace)
    {
        _handler.store(handler, std::memory_order_release);
    }
    return _isHandlerInPlace;
}
```

```117:134:profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/TimerCreateCpuProfiler.cpp
    auto registered = _pSignalManager->RegisterHandler(TimerCreateCpuProfiler::CollectStackSampleSignalHandler);

    if (registered)
    {
        // ...
    }

    return registered;
```

**Why it matters:** The guard looks armed (`Run` always `sigsetjmp`s) even when SIGSEGV/SIGBUS were never hooked. A later dump-time fault then kills the process instead of skipping one object/type. `call_once` also means a transient `sigaction` failure is permanent for the process.

**What to do:** Capture both `RegisterHandler` results. On failure, log `Error` with the signal name. Have `Run()` on Linux no-op / return true without arming TLS if install failed — better to skip recovery than pretend it exists. Optionally retry on the next `EnsureInstalled` if the first `sigaction` failed (don’t wrap a failed install in `call_once` that never retries). Match `TimerCreateCpuProfiler`: treat `false` as “this feature is not available.”

---

## 3. Fault during `ResolvePendingTypes` skips COM/C++ destructors (Low)

**Today:** After the dump, each pending `ClassID` is inspected under `MemoryFaultGuard::Run` wrapping the **entire** `GetOrBuildInlineVTInfo` → `BuildInlineVTInfo`. That path owns `ComPtr<IMetaDataImport>` (destructor calls `Release`), plus `std::vector`s (`typeArgs`, `inlineVtFields`, `fieldOffsets`). A SIGSEGV/`siglongjmp` (Linux) or SEH (Windows) skips those destructors.

The header already says `GetClassIDInfo2` is the intended safe validation (runtime error code for a dead ClassID); the raw MethodTable read is `GCDesc::ContainsGCPointers`. The guard comment admits the whole inspection is wrapped “because the metadata calls take the same stale pointer.” The guard’s own contract says neither the body nor anything it calls may own something that needs destruction.

```116:134:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
    for (ClassID classID : pendingClassIDs)
    {
        // The whole inspection is guarded, not just the
        // MethodTable read, because the metadata calls take the same stale pointer.
        if (!MemoryFaultGuard::Run([this, classID] { GetOrBuildInlineVTInfo(classID); }))
        {
            faultedCount++;
            _cache.insert_or_assign(classID, std::nullopt);
        }
    }
```

```229:264:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
std::optional<InlineVTCache::InlineVTInfo> InlineVTCache::BuildInlineVTInfo(ClassID classID)
{
    ModuleID moduleID = 0;
    mdTypeDef typeDef = mdTokenNil;
    ClassID parentClassID = 0;
    std::vector<ClassID> typeArgs;
    ComPtr<IMetaDataImport> pMetadataImport;

    // Deliberately the first thing done with the type: GetClassIDInfo2 (called from here)
    // is validated by the runtime and fails with an error code for a ClassID it no longer
    // knows, where the MethodTable read below would fault on freed memory.
    if (!TryGetTypeDefAndMetadata(classID, moduleID, typeDef, pMetadataImport, &parentClassID, &typeArgs))
    {
        return std::nullopt;
    }

    if (!GCDesc::ContainsGCPointers(classID))
    {
        return std::nullopt;
    }
```

```145:149:profiler/src/ProfilerEngine/Datadog.Profiler.Native/MemoryFaultGuard.h
// IMPORTANT: neither body nor anything it calls may own something that needs
// destruction, because nothing releases it on the fault path: SEH unwinding skips the
// destructors of intervening frames and siglongjmp does not unwind at all.
```

**Why it matters:** A fault mid-inspection leaks the metadata `IUnknown` and the vectors. Unlikely, but it violates the guard’s own contract. `GetClassIDInfo2` / `GetModuleMetaData` are CLR APIs, not raw pointer derefs; the thing that actually SIGSEGVs is the MethodTable read.

**What to do:** Narrow the guard to the raw MethodTable access only. Call `TryGetTypeDefAndMetadata` (GetClassIDInfo2 + GetModuleMetaData) **outside** `Run`. Then:

```cpp
bool hasGCPointers = false;
if (!MemoryFaultGuard::Run([&] { hasGCPointers = GCDesc::ContainsGCPointers(classID); }))
{
    // cache nullopt, continue
}
if (!hasGCPointers) { /* cache nullopt */ }
// rest of BuildInlineVTInfo unguarded
```

If you still want a belt for the unload-during-loop race on later CLR calls, catch C++ exceptions around the unguarded remainder rather than `siglongjmp`ing through `ComPtr`. Keep `GetClassIDInfo2` as the first, unguarded validation — that part of the design is correct.

---

## 4. Dead `_pFrameStore` on `InlineVTCache` (Nit)

**Today:** The constructor stores `pFrameStore` in `_pFrameStore`. Nothing in `InlineVTCache.cpp` / `.h` ever reads it. Tests already construct with `nullptr`. `HeapSnapshotManager` still passes a real `IFrameStore*` that is unused here (it uses its own `_pFrameStore` for histogram names).

```11:21:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
InlineVTCache::InlineVTCache(
    ICorProfilerInfo12* pCorProfilerInfo,
    IFrameStore* pFrameStore,
    CoreLibModuleProvider* pCoreLibModuleProvider)
    :
    _pCorProfilerInfo(pCorProfilerInfo),
    _pFrameStore(pFrameStore),
    _pCoreLibModuleProvider(pCoreLibModuleProvider)
{
    MemoryFaultGuard::EnsureInstalled();
}
```

```205:207:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.h
    ICorProfilerInfo12* _pCorProfilerInfo;
    IFrameStore* _pFrameStore;
    CoreLibModuleProvider* _pCoreLibModuleProvider;
```

Grep of `InlineVTCache.cpp`: `_pFrameStore` appears only on the initializer line. Contrast `ReferenceChainTraverser`, which actually calls `_pFrameStore->GetTypeName`.

**Why it matters:** Dead field/parameter; every construction site (manager + tests) threads an unused pointer.

**What to do:** Remove `_pFrameStore` and the ctor parameter. Update `HeapSnapshotManager.cpp:94`, `InlineVTCacheTest.cpp` helper, and every `InlineVTCache(pInfo, &frameStore, …)` in `ReferenceChainTraverserFaultTest.cpp` to drop the middle arg. Do not keep it “for later” unless there is a concrete name-resolution use.

---

## 5. Signature parsing can read past the blob (Nit)

**Today:** Several paths do `idx++` then `CorSigUncompressToken(&pSignature[idx], …)` without checking that enough bytes remain. Compressed tokens are 1–4 bytes. The length-unaware overload in `cor.h` does not take a remaining length; it will read past `signatureSize`. The CMOD loop and `ResolveGenericArgClassID`’s `ELEMENT_TYPE_VALUETYPE`/`CLASS` path have **no** check after `idx++`. The `ELEMENT_TYPE_VALUETYPE` field path and GENERICINST path check `idx >= signatureSize` (one byte) but not the 1–4 bytes `CorSigUncompressToken` may consume.

```328:336:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
            if (prefix == ELEMENT_TYPE_CMOD_OPT || prefix == ELEMENT_TYPE_CMOD_REQD)
            {
                idx++;
                mdToken token;
                idx += CorSigUncompressToken(&pSignature[idx], &token);
                continue;
            }
```

```373:381:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
        else if (elementType == ELEMENT_TYPE_VALUETYPE)
        {
            idx++;
            if (idx >= signatureSize)
            {
                return;
            }
            mdToken vtToken;
            idx += CorSigUncompressToken(&pSignature[idx], &vtToken);
```

```437:441:profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp
        if (argType == ELEMENT_TYPE_VALUETYPE || argType == ELEMENT_TYPE_CLASS)
        {
            idx++;
            mdToken argToken = mdTokenNil;
            idx += CorSigUncompressToken(&pSignature[idx], &argToken);
```

Length-aware API already in-tree:

```1990:1994:shared/src/native-lib/coreclr/src/inc/cor.h
inline HRESULT CorSigUncompressToken(
    PCCOR_SIGNATURE pData,
    uint32_t dwLen,
    mdToken * pToken,
    uint32_t * dwTokenLength)
```

**Why it matters:** A truncated or malformed field signature (dynamic modules, corrupt metadata) can read off the end of the blob. This is not a heap-object walk, but it is still an OOB read in the profiler.

**What to do:** Use the length-aware overload at every token/data uncompress. Helper:

```cpp
static bool TryUncompressToken(PCCOR_SIGNATURE pSignature, ULONG signatureSize, ULONG& idx, mdToken& token)
{
    if (idx >= signatureSize) return false;
    uint32_t consumed = 0;
    if (FAILED(CorSigUncompressToken(&pSignature[idx], signatureSize - idx, &token, &consumed)))
        return false;
    idx += consumed;
    return true;
}
```

Same pattern for `CorSigUncompressData` (`ELEMENT_TYPE_VAR`, nested arg count). On failure, return 0 / leave `fieldInfo` unset. Apply in the CMOD loop, both `ELEMENT_TYPE_VALUETYPE` paths, and GENERICINST.

---

## 6. PR description / histogram TODO (Nit)

**Today:** PR 8966 body still says:

> - Add namespace to type name in reference chain
> - TODO: avoid stopping the generation of class histogram on error

The histogram already continues when traversal fails. `OnBulkNodes` always fills `_classHistogram` and never consults the traverser. `_gcDescTrusted` / `_gcDescDisabled` only skip creating or running `ReferenceChainTraverser`. Comments on both types say the class histogram is unaffected.

```303:330:profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp
void HeapSnapshotManager::OnBulkNodes(
    uint32_t index,
    uint32_t count,
    GCBulkNodeValue* pNodes)
{
    std::lock_guard lock(_histogramLock);
    _objectCount += count;
    for (size_t i = 0; i < count; i++)
    {
        // ... updates _classHistogram from pNodes[i].TypeID / Size
        // no traverser / _gcDescTrusted check
```

```686:697:profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp
    // If the GCDesc reader previously failed its self-test, do not create the
    // traverser: the reference tree is skipped while the class histogram still runs.
    if (_gcDescDisabled)
    {
        _pReferenceChainTraverser.reset();
    }
    else
    {
        _pReferenceChainTraverser = std::make_unique<ReferenceChainTraverser>(
            _pCorProfilerInfo, _pFrameStore, *_typeReferenceTree, *_pInlineVTCache,
            _visitedSetHighWatermark);
    }
```

```238:241:profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.h
    // Set to true once the GCDesc reader fails its runtime self-test during a dump,
    // or once memory access faults exhaust the budget on several consecutive dumps.
    // When set, subsequent dumps skip reference-chain traversal entirely (no
    // traverser is created) while the class histogram continues to work.
```

**Why it matters:** Reviewers still think histogram-on-error is an open TODO. It is already implemented: traversal disable is isolated; histogram is a separate `OnBulkNodes` path.

**What to do:** Edit the PR body. Drop the TODO. Summary should mention: namespaces on type names; fault guard + per-dump fault budget; histogram **already** continues when traversal is aborted or permanently disabled (`_gcDescDisabled` only skips the traverser). Implementation details should mention serialization namespaces plus the guard/disable policy, not a pending histogram fix.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp (hunk 3 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/InlineVTCache.cpp:71]: Does not seem to be an issue

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/HeapSnapshotManager.cpp:121]: Yes but in a later PR with the other services starting threads in the profiler

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp (hunk 2 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp]: ✅ done

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp (hunk 1 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/CoreLibModuleProvider.cpp]: ✅ done

**Comment by chrisnas** [comment]
The issues found by the AI have been fixed
